db/bigtable.go

```diff
@@ -1016,157 +1287,172 @@ func (bigtable *Bigtable) GetValidatorMissedAttestationHistory(validators []uint
 
-	if len(columnFilters) == 1 { // special case to retrieve data for one validators
-		filter = gcp_bigtable.ChainFilters(
-			gcp_bigtable.FamilyFilter(ATTESTATIONS_FAMILY),
-			columnFilters[0],
-		)
-	}
-	if len(columnFilters) == 0 { // special case to retrieve data for all validators
-		filter = gcp_bigtable.FamilyFilter(ATTESTATIONS_FAMILY)
-	}
+	g, gCtx := errgroup.WithContext(ctx)
+	g.SetLimit(concurrency)
 
-	err = bigtable.tableBeaconchain.ReadRows(ctx, ranges, func(r gcp_bigtable.Row) bool {
-		keySplit := strings.Split(r.Key(), ":")
+	for i := 0; i < len(validators); i += batchSize {
 
-		attesterSlot, err := strconv.ParseUint(keySplit[4], 10, 64)
-		if err != nil {
-			logger.Errorf("error parsing slot from row key %v: %v", r.Key(), err)
-			return false
+		upperBound := i + batchSize
+		if len(validators) < upperBound {
+			upperBound = len(validators)
 		}
-		attesterSlot = max_block_number - attesterSlot
+		vals := validators[i:upperBound]
 
-		for _, ri := range r[ATTESTATIONS_FAMILY] {
-			inclusionSlot := max_block_number - uint64(ri.Timestamp)/1000
-
-			status := uint64(1)
-			if inclusionSlot == max_block_number {
-				status = 0
-			}
-
-			validator, err := strconv.ParseUint(strings.TrimPrefix(ri.Column, ATTESTATIONS_FAMILY+":"), 10, 64)
-			if err != nil {
-				logger.Errorf("error parsing validator from column key %v: %v", ri.Column, err)
-				return false
+		g.Go(func() error {
+			select {
+			case <-gCtx.Done():
+				return gCtx.Err()
+			default:
 			}
+			ranges := bigtable.getValidatorsEpochRanges(vals, startEpoch, endEpoch)
+			err = bigtable.tableValidatorAttestations.ReadRows(ctx, ranges, func(r gcp_bigtable.Row) bool {
+				keySplit := strings.Split(r.Key(), ":")
 
-			// only if the attestation was not included in another slot we count it as missed
-			if (status == 0 || orphanedSlotsMap[inclusionSlot]) && (foundValid[validator] == nil || !foundValid[validator][attesterSlot]) {
-				if res[validator] == nil {
-					res[validator] = make(map[uint64]bool, 0)
-				}
-				res[validator][attesterSlot] = true
-			} else {
-				if res[validator] != nil && res[validator][attesterSlot] {
-					delete(res[validator], attesterSlot)
+				validator, err := bigtable.validatorKeyToIndex(keySplit[1])
+				if err != nil {
+					logger.Errorf("error parsing validator from row key %v: %v", r.Key(), err)
+					return false
 				}
-				if foundValid[validator] == nil {
-					foundValid[validator] = make(map[uint64]bool, 0)
+
+				for _, ri := range r[ATTESTATIONS_FAMILY] {
+					attesterSlotString := strings.Replace(ri.Column, ATTESTATIONS_FAMILY+":", "", 1)
+					attesterSlot, err := strconv.ParseUint(attesterSlotString, 10, 64)
+					if err != nil {
+						logger.Errorf("error parsing slot from row key %v: %v", r.Key(), err)
+						return false
+					}
+
+					inclusionSlot := MAX_BLOCK_NUMBER - uint64(ri.Timestamp)/1000
+
+					status := uint64(1)
+					if inclusionSlot == MAX_BLOCK_NUMBER {
```

Comment:
not sure if this is necessary. Same as with the other occurence of this, inclusionSlot should be 0 if you want to handle the 0 case. As for the MAX_BLOCK_NUMBER slot case, not sure if setting the status for this future slot is necessary